pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/text/NativeTextIndexReader.java

```diff
@@ -0,0 +1,133 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class NativeTextIndexReader implements TextIndexReader {
+  private static final Logger LOGGER = LoggerFactory.getLogger(NativeTextIndexReader.class);
+
+  private final String _column;
+  private final File _indexFile;
+  private final PinotDataBuffer _buffer;
+
+  private FST _fst;
+  private BitmapInvertedIndexReader _invertedIndex;
+
+  public NativeTextIndexReader(String column, File indexDir) {
+    _column = column;
+    try {
+      _indexFile = getTextIndexFile(indexDir);
+      _buffer = PinotDataBuffer.loadBigEndianFile(_indexFile);
+
+      populateIndexes();
+    } catch (Exception e) {
+      LOGGER.error("Failed to instantiate Lucene text index reader for column {}, exception {}", column,
+          e.getMessage());
+      throw new RuntimeException(e);
+    }
+  }
+
+  private File getTextIndexFile(File segmentIndexDir) {
+    // will return null if file does not exist
+    File file = SegmentDirectoryPaths.findTextIndexIndexFileNative(segmentIndexDir, _column);
+    if (file == null) {
+      throw new IllegalStateException("Failed to find text index file for column: " + _column);
+    }
+    return file;
+  }
+
+  private void populateIndexes() {
+    int fstMagic = _buffer.getInt(0);
+
+    if (fstMagic != FSTHeader.FST_MAGIC) {
+      throw new IllegalStateException("Native FST declared but header is not matching");
+    }
+
+    long invertedIndexLength = _buffer.getLong(4);
+    long fstDataLength = _buffer.getLong(12);
+    int numBitMaps = _buffer.getInt(20);
+
+    long fstDataStartOffset = NativeTextIndexCreator.HEADER_LENGTH;
+    long fstDataEndOffset = fstDataStartOffset + fstDataLength;
+    ByteBuffer byteBuffer = _buffer.toDirectByteBuffer(fstDataStartOffset, (int) fstDataLength);
+    try {
+      _fst = FST.read(new ByteBufferInputStream(Collections.singletonList(byteBuffer)), ImmutableFST.class, true);
+    } catch (IOException e) {
+      throw new RuntimeException(e.getMessage());
+    }
+
+    long invertedIndexEndOffset = fstDataEndOffset + invertedIndexLength;
+    _invertedIndex =
+        new BitmapInvertedIndexReader(_buffer.view(fstDataEndOffset, invertedIndexEndOffset, ByteOrder.BIG_ENDIAN),
+            numBitMaps);
+  }
+
+  @Override
+  public ImmutableRoaringBitmap getDictIds(String searchQuery) {
+    throw new UnsupportedOperationException("");
+  }
+
+  @Override
+  public MutableRoaringBitmap getDocIds(String searchQuery) {
+    try {
+      List<Integer> returnList = new ArrayList<>();
+      RegexpMatcher.regexMatch(searchQuery, _fst, returnList);
+      MutableRoaringBitmap matchingDocIds = null;
+
+      for (int dictId : returnList) {
+        ImmutableRoaringBitmap docIds = _invertedIndex.getDocIds(dictId);
+        if (matchingDocIds == null) {
+          matchingDocIds = docIds.toMutableRoaringBitmap();
+        } else {
+          matchingDocIds.or(docIds);
+        }
+      }
```

Comment:
(optional) It will be faster to collect the bitmaps into an array and or them afterwards, because it avoids computing the cardinalities of each intermediate union.

Reply:
Will do that as a follow up